src/cmclient/cmclient.cpp

```diff
@@ -39,49 +44,20 @@ aos::Error CMClient::Init(LauncherItf& launcher, ResourceManager& resourceManage
     mResourceManager = &resourceManager;
     mDownloader = &downloader;
 
-    auto err = mThread.Run([this](void*) {
-        while (true) {
-            ConnectToCM();
-
-            atomic_set_bit(&mState, eConnected);
-
-            auto err = SendNodeConfiguration();
-            if (!err.IsNone()) {
-                LOG_ERR() << "Can't send node configuration: " << err;
-            }
-
-            err = mLauncher->RunLastInstances();
-            if (!err.IsNone()) {
-                LOG_ERR() << "Can't run last instances: " << err;
-            }
-
-            err = ProcessMessages();
-
-            vch_close(&mSMVChanHandler);
-            atomic_clear_bit(&mState, eConnected);
-
-            if (!err.IsNone()) {
-                LOG_ERR() << "Error processing messages: " << err;
-
-                continue;
-            }
-
-            break;
-        }
-    });
-    if (!err.IsNone()) {
-        return AOS_ERROR_WRAP(err);
+    auto ret = RunWriter();
+    if (!ret.IsNone()) {
+        return ret;
     }
 
-    return aos::ErrorEnum::eNone;
+    return RunReader();
 }
 
 aos::Error CMClient::InstancesRunStatus(const aos::Array<aos::InstanceStatus>& instances)
 {
     aos::LockGuard lock(mMutex);
 
-    if (!atomic_test_bit(&mState, eConnected)) {
-        return AOS_ERROR_WRAP(aos::ErrorEnum::eWrongState);
+    if (atomic_test_bit(&mState, eFinish)) {
```

Comment:
Could we move this condition inside `SendBufferToVChan`?

Reply:
Agree, it is much better.